File: implementations/react/src/patterns/behavioral/visitor/visitor-total-price.ts
import type { Book } from "./element-book";
import type { Electronics } from "./element-electronics";
import type { Grocery } from "./element-grocery";
import type { Visitor } from "./visitor";

/**
 * Concrete Visitor que calcula el precio total con impuestos
 */

export class TotalPriceVisitor implements Visitor {
  subtotal = 0;
  tax = 0;
  total = 0;

  visitBook(e: Book) {
    const base = e.unitPrice * e.qty;
    const tax = (e.isImported ? 0.05 : 0) * base;   
    this.subtotal += base; this.tax += tax; this.total += base + tax;
  }

  visitElectronics(e: Electronics) {
    const base = e.unitPrice * e.qty;
    const tax = 0.15 * base;                    
    this.subtotal += base; this.tax += tax; this.total += base + tax;
  }

  visitGrocery(e: Grocery) {
    const base = e.unitPrice * e.qty;
    const tax = 0;                                  
    this.subtotal += base; this.tax += tax; this.total += base + tax;
  }
}

File: implementations/flutter/design_patterns_flutter/lib/patterns/behavioral/visitor/visitor_total_price.dart
import 'visitor.dart';
import 'element_book.dart';
import 'element_electronics.dart';
import 'element_grocery.dart';

class TotalPriceVisitor implements Visitor {
  double subtotal = 0;
  double tax = 0;
  double total = 0;

  @override
  void visitBook(Book e) {
    final base = e.unitPrice * e.qty;
    final t = (e.isImported ? 0.05 : 0) * base; // books exempt, only import
    subtotal += base;
    tax += t;
    total += base + t;
  }

  @override
  void visitElectronics(Electronics e) {
    final base = e.unitPrice * e.qty;
    final t = 0.15 * base;
    subtotal += base;
    tax += t;
    total += base + t;
  }

  @override
  void visitGrocery(Grocery e) {
    final base = e.unitPrice * e.qty;
    final t = 0; // groceries taxed 0
    subtotal += base;
    tax += t;
    total += base + t;
  }
}
